File: substrate/floating_dr_draft_before_merge_mo_price_ceiling_compose.py
"""Floating DR highlight over draft-before-merge MO price-ceiling recursive twin (pure).

live_dispatched / merge_executed / pack_dispatched / twin_written always False.
draft_written / live_execution_authorized / charge_executed always False.
production_router_verdict always REJECT.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from substrate.draft_before_merge_mo_price_ceiling_recursive_twin_compose import (
    DraftBeforeMergeMoPriceCeilingRecursiveTwinCompose,
    DraftBeforeMergeMoPriceCeilingRecursiveTwinComposeError,
    compose_draft_before_merge_mo_price_ceiling_recursive_twin,
)
from substrate.reading_highlight_float_twin_feed_compose import (
    ReadingHighlightFloatTwinFeedCompose,
    ReadingHighlightFloatTwinFeedComposeError,
    compose_reading_highlight_float_twin_feed,
)


class FloatingDrDraftBeforeMergeMoPriceCeilingComposeError(ValueError):
    """Fail-closed validation for floating DR + draft-before-merge MO pack."""


@dataclass(frozen=True)
class FloatingDrDraftBeforeMergeMoPriceCeilingCompose:
    session_id: str
    parent_asset_id: str
    week_id: str
    asset_id: str
    title: str
    account_id: str
    highlight_surface: ReadingHighlightFloatTwinFeedCompose
    draft_pack: DraftBeforeMergeMoPriceCeilingRecursiveTwinCompose
    session_aligned: bool
    parent_aligned: bool
    pack_ready: bool
    live_dispatched: bool
    merge_executed: bool
    pack_dispatched: bool
    twin_written: bool
    record_persisted: bool
    draft_written: bool
    live_execution_authorized: bool
    charge_executed: bool
    prompts_injected: bool
    live_router_authorized: bool
    secrets_stored: bool
    live_meter_read: bool
    remote_index_queried: bool
    backlog_mutated: bool
    store_mutated: bool
    suite_rewritten: bool
    pdf_view_authorized: bool
    pdf_primary: bool
    live_dispatch_authorized: bool
    inventory_mutated: bool
    analysis_written: bool
    purchase_executed: bool
    hosted: bool
    remote_fetched: bool
    production_router_verdict: str
    notes: tuple[str, ...]
    authority: str

    def to_dict(self) -> dict[str, Any]:
        return {
            "session_id": self.session_id,
            "parent_asset_id": self.parent_asset_id,
            "week_id": self.week_id,
            "asset_id": self.asset_id,
            "title": self.title,
            "account_id": self.account_id,
            "highlight_surface": self.highlight_surface.to_dict(),
            "draft_pack": self.draft_pack.to_dict(),
            "session_aligned": self.session_aligned,
            "parent_aligned": self.parent_aligned,
            "pack_ready": self.pack_ready,
            "live_dispatched": False,
            "merge_executed": False,
            "pack_dispatched": False,
            "twin_written": False,
            "record_persisted": False,
            "draft_written": False,
            "live_execution_authorized": False,
            "charge_executed": False,
            "prompts_injected": False,
            "live_router_authorized": False,
            "secrets_stored": False,
            "live_meter_read": False,
            "remote_index_queried": False,
            "backlog_mutated": False,
            "store_mutated": False,
            "suite_rewritten": False,
            "pdf_view_authorized": False,
            "pdf_primary": False,
            "live_dispatch_authorized": False,
            "inventory_mutated": False,
            "analysis_written": False,
            "purchase_executed": False,
            "hosted": False,
            "remote_fetched": False,
            "production_router_verdict": "REJECT",
            "notes": list(self.notes),
            "authority": (
                "floating_dr_draft_before_merge_mo_price_ceiling_compose_advisory"
            ),
        }


def _require_nonempty(value: object, *, field: str) -> str:
    if not isinstance(value, str) or not value.strip():
        raise FloatingDrDraftBeforeMergeMoPriceCeilingComposeError(
            f"{field} must be a non-empty string"
        )
    return value.strip()


def compose_floating_dr_draft_before_merge_mo_price_ceiling(
    *,
    highlight_surface: object,
    draft_pack: object,
    operator_ack: object,
    require_both: object | None = None,
) -> FloatingDrDraftBeforeMergeMoPriceCeilingCompose:
    """Highlight floating DR on draft-before-merge MO pack. Never dispatches."""
    if not isinstance(operator_ack, bool):
        raise FloatingDrDraftBeforeMergeMoPriceCeilingComposeError(
            "operator_ack must be an explicit boolean"
        )
    if not isinstance(highlight_surface, dict):
        raise FloatingDrDraftBeforeMergeMoPriceCeilingComposeError(
            "highlight_surface must be an object"
        )
    if not isinstance(draft_pack, dict):
        raise FloatingDrDraftBeforeMergeMoPriceCeilingComposeError(
            "draft_pack must be an object"
        )

    require = True if require_both is None else require_both
    if not isinstance(require, bool):
        raise FloatingDrDraftBeforeMergeMoPriceCeilingComposeError(
            "require_both must be boolean when set"
        )

    notes: list[str] = [
        "live_dispatched=false · merge_executed=false · pack_dispatched=false",
        "twin_written=false · draft_written=false · live_execution_authorized=false",
        "production_router_verdict=REJECT",
    ]

    try:
        hs = compose_reading_highlight_float_twin_feed(
            session_id=highlight_surface.get("session_id"),
            parent_asset_id=highlight_surface.get("parent_asset_id"),
            highlight=highlight_surface.get("highlight"),
            gated=highlight_surface.get("gated"),
            would_exceed=highlight_surface.get("would_exceed"),
            surface_action=highlight_surface.get("surface_action"),
            operator_ack=operator_ack,
            prompt=highlight_surface.get("prompt"),
            preferred_view_mode=highlight_surface.get("preferred_view_mode"),
            operator_override=highlight_surface.get("operator_override"),
            selected_model_id=highlight_surface.get("selected_model_id"),
            source_families=highlight_surface.get("source_families"),
            existing_members=highlight_surface.get("existing_members"),
            selected_instance_ids=highlight_surface.get("selected_instance_ids"),
            twin_findings=highlight_surface.get("twin_findings"),
            existing_twin_asset_id=highlight_surface.get(
                "existing_twin_asset_id"
            ),
            mark_for_prompt_context=highlight_surface.get(
                "mark_for_prompt_context"
            ),
            include_twin_feed=highlight_surface.get("include_twin_feed"),
        )
    except ReadingHighlightFloatTwinFeedComposeError as e:
        raise FloatingDrDraftBeforeMergeMoPriceCeilingComposeError(
            str(e)
        ) from e
    notes.extend(f"[highlight_surface] {n}" for n in hs.notes)

    try:
        dp = compose_draft_before_merge_mo_price_ceiling_recursive_twin(
            draft_gate=draft_pack.get("draft_gate"),
            mo_pack=draft_pack.get("mo_pack"),
            operator_ack=operator_ack,
            require_both=draft_pack.get("require_both"),
        )
    except DraftBeforeMergeMoPriceCeilingRecursiveTwinComposeError as e:
        raise FloatingDrDraftBeforeMergeMoPriceCeilingComposeError(
            str(e)
        ) from e
    notes.extend(f"[draft_pack] {n}" for n in dp.notes)

    session = _require_nonempty(hs.session_id, field="session_id")
    parent = _require_nonempty(
        highlight_surface.get("parent_asset_id"), field="parent_asset_id"
    )
    week = _require_nonempty(dp.week_id, field="week_id")
    asset = _require_nonempty(dp.asset_id, field="asset_id")
    title = _require_nonempty(dp.title, field="title")
    account = _require_nonempty(dp.account_id, field="account_id")

    session_aligned = dp.session_id == session
    parent_aligned = dp.parent_asset_id == parent or dp.asset_id == parent
    if not session_aligned:
        notes.append(
            "session_id mismatch between highlight_surface and draft_pack — "
            "pack_ready blocked"
        )
    else:
        notes.append("session_aligned=true")
    if not parent_aligned:
        notes.append(
            "parent_asset_id mismatch between highlight_surface and draft_pack "
            "— pack_ready blocked"
        )
    else:
        notes.append("parent_aligned=true")

    if require:
        pack_ready = (
            session_aligned is True
            and parent_aligned is True
            and hs.pack_ready is True
            and dp.pack_ready is True
            and dp.production_router_verdict == "REJECT"
            and hs.live_dispatched is False
            and hs.merge_executed is False
            and hs.pack_dispatched is False
            and hs.twin_written is False
            and dp.draft_written is False
            and dp.merge_executed is False
            and dp.live_dispatched is False
            and dp.live_execution_authorized is False
            and dp.charge_executed is False
            and dp.twin_written is False
            and dp.remote_index_queried is False
            and dp.pdf_primary is False
            and dp.live_router_authorized is False
            and dp.secrets_stored is False
            and operator_ack is True
        )
    else:
        pack_ready = (
            session_aligned is True
            and parent_aligned is True
            and operator_ack is True
            and dp.production_router_verdict == "REJECT"
            and dp.pdf_primary is False
            and hs.live_dispatched is False
            and (hs.pack_ready is True or dp.pack_ready is True)
        )

    if pack_ready:
        notes.append(
            "pack_ready=true — floating DR highlight + draft-before-merge MO "
            "price-ceiling ready; still pure"
        )
    else:
        notes.append(
            "pack_ready=false — highlight_surface, draft_pack, alignment, or "
            "operator_ack gate open"
        )

    if (
        hs.live_dispatched is not False
        or hs.merge_executed is not False
        or hs.pack_dispatched is not False
        or hs.twin_written is not False
        or dp.draft_written is not False
        or dp.merge_executed is not False
        or dp.live_dispatched is not False
        or dp.live_execution_authorized is not False
        or dp.charge_executed is not False
        or dp.twin_written is not False
        or dp.remote_index_queried is not False
        or dp.pdf_primary is not False
        or dp.live_router_authorized is not False
        or dp.secrets_stored is not False
        or dp.production_router_verdict != "REJECT"
    ):
        raise FloatingDrDraftBeforeMergeMoPriceCeilingComposeError(
            "invariant: honesty flags must remain false / REJECT"
        )

    notes.extend(
        (
            "live_dispatched=false",
            "merge_executed=false",
            "pack_dispatched=false",
            "twin_written=false",
            "record_persisted=false",
            "draft_written=false",
            "live_execution_authorized=false",
            "charge_executed=false",
            "prompts_injected=false",
            "live_router_authorized=false",
            "secrets_stored=false",
            "live_meter_read=false",
            "remote_index_queried=false",
            "backlog_mutated=false",
            "store_mutated=false",
            "suite_rewritten=false",
            "pdf_view_authorized=false",
            "pdf_primary=false",
            "live_dispatch_authorized=false",
            "inventory_mutated=false",
            "analysis_written=false",
            "purchase_executed=false",
            "hosted=false",
            "remote_fetched=false",
            "production_router_verdict=REJECT",
        )
    )

    return FloatingDrDraftBeforeMergeMoPriceCeilingCompose(
        session_id=session,
        parent_asset_id=parent,
        week_id=week,
        asset_id=asset,
        title=title,
        account_id=account,
        highlight_surface=hs,
        draft_pack=dp,
        session_aligned=session_aligned,
        parent_aligned=parent_aligned,
        pack_ready=pack_ready,
        live_dispatched=False,
        merge_executed=False,
        pack_dispatched=False,
        twin_written=False,
        record_persisted=False,
        draft_written=False,
        live_execution_authorized=False,
        charge_executed=False,
        prompts_injected=False,
        live_router_authorized=False,
        secrets_stored=False,
        live_meter_read=False,
        remote_index_queried=False,
        backlog_mutated=False,
        store_mutated=False,
        suite_rewritten=False,
        pdf_view_authorized=False,
        pdf_primary=False,
        live_dispatch_authorized=False,
        inventory_mutated=False,
        analysis_written=False,
        purchase_executed=False,
        hosted=False,
        remote_fetched=False,
        production_router_verdict="REJECT",
        notes=tuple(notes),
        authority=(
            "floating_dr_draft_before_merge_mo_price_ceiling_compose_advisory"
        ),
    )


def format_floating_dr_draft_before_merge_mo_price_ceiling_summary(
    c: FloatingDrDraftBeforeMergeMoPriceCeilingCompose,
) -> str:
    return (
        f"pack_ready={c.pack_ready} · "
        f"highlight_ready={c.highlight_surface.pack_ready} · "
        f"draft_ready={c.draft_pack.pack_ready} · "
        f"session_aligned={c.session_aligned} · "
        f"parent_aligned={c.parent_aligned} · "
        f"verdict={c.production_router_verdict} · "
        "live_dispatched=false · twin_written=false · draft_written=false · "
        "live_execution_authorized=false"
    )


__all__ = [
    "FloatingDrDraftBeforeMergeMoPriceCeilingCompose",
    "FloatingDrDraftBeforeMergeMoPriceCeilingComposeError",
    "compose_floating_dr_draft_before_merge_mo_price_ceiling",
    "format_floating_dr_draft_before_merge_mo_price_ceiling_summary",
]

File: apps/reading/src/api/floatingDrDraftBeforeMergeMoPriceCeilingCompose.ts
/**
 * Reading highlight floating deep-research residual over draft-before-merge +
 * MO price-ceiling + recursive twin note-taker pack (pure).
 *
 * Operator vision: from a reading highlight, spin floating deep research /
 * tray + twin feed, while draft-before-full-merge, Midnight Oil price ceiling,
 * and recursive twin honesty remain pure — never live-dispatches, merges
 * parent, launches MO, or writes twins.
 *
 * live_dispatched / merge_executed / pack_dispatched / twin_written always false.
 * live_execution_authorized / charge_executed / draft_written always false.
 * production_router_verdict always REJECT.
 */

import {
  composeReadingHighlightFloatTwinFeed,
  type ReadingHighlightFloatTwinFeedCompose,
  type ReadingHighlightFloatTwinFeedInput,
} from "./readingHighlightFloatTwinFeedCompose";
import {
  composeDraftBeforeMergeMoPriceCeilingRecursiveTwin,
  type DraftBeforeMergeMoPriceCeilingRecursiveTwinCompose,
  type DraftBeforeMergeMoPriceCeilingRecursiveTwinInput,
} from "./draftBeforeMergeMoPriceCeilingRecursiveTwinCompose";

export interface FloatingDrDraftBeforeMergeMoPriceCeilingInput {
  highlight_surface: Omit<ReadingHighlightFloatTwinFeedInput, "operator_ack">;
  draft_pack: Omit<
    DraftBeforeMergeMoPriceCeilingRecursiveTwinInput,
    "operator_ack"
  >;
  operator_ack: boolean;
  /**
   * When true (default), require highlight_surface.pack_ready AND
   * draft_pack.pack_ready, plus session/parent alignment.
   */
  require_both?: boolean;
}

export interface FloatingDrDraftBeforeMergeMoPriceCeilingCompose {
  session_id: string;
  parent_asset_id: string;
  week_id: string;
  asset_id: string;
  title: string;
  account_id: string;
  highlight_surface: ReadingHighlightFloatTwinFeedCompose;
  draft_pack: DraftBeforeMergeMoPriceCeilingRecursiveTwinCompose;
  session_aligned: boolean;
  parent_aligned: boolean;
  pack_ready: boolean;
  live_dispatched: false;
  merge_executed: false;
  pack_dispatched: false;
  twin_written: false;
  record_persisted: false;
  draft_written: false;
  live_execution_authorized: false;
  charge_executed: false;
  prompts_injected: false;
  live_router_authorized: false;
  secrets_stored: false;
  live_meter_read: false;
  remote_index_queried: false;
  backlog_mutated: false;
  store_mutated: false;
  suite_rewritten: false;
  pdf_view_authorized: false;
  pdf_primary: false;
  live_dispatch_authorized: false;
  inventory_mutated: false;
  analysis_written: false;
  purchase_executed: false;
  hosted: false;
  remote_fetched: false;
  production_router_verdict: "REJECT";
  notes: string[];
  authority: "floating_dr_draft_before_merge_mo_price_ceiling_compose_advisory";
}

function requireNonEmpty(value: unknown, name: string): string {
  if (typeof value !== "string" || !value.trim()) {
    throw new Error(`${name} must be a non-empty string`);
  }
  return value.trim();
}

/**
 * Highlight floating DR stacked on draft-before-merge MO price-ceiling recursive twin.
 * Never live-dispatches; never merges; ND REJECT.
 */
export function composeFloatingDrDraftBeforeMergeMoPriceCeiling(
  input: FloatingDrDraftBeforeMergeMoPriceCeilingInput,
): FloatingDrDraftBeforeMergeMoPriceCeilingCompose {
  if (!input || typeof input !== "object") {
    throw new Error("input must be an object");
  }
  if (typeof input.operator_ack !== "boolean") {
    throw new Error("operator_ack must be an explicit boolean");
  }
  if (!input.highlight_surface || typeof input.highlight_surface !== "object") {
    throw new Error("highlight_surface must be an object");
  }
  if (!input.draft_pack || typeof input.draft_pack !== "object") {
    throw new Error("draft_pack must be an object");
  }

  const require_both =
    input.require_both === undefined ? true : input.require_both;
  if (typeof require_both !== "boolean") {
    throw new Error("require_both must be boolean when set");
  }

  const notes: string[] = [
    "live_dispatched=false · merge_executed=false · pack_dispatched=false",
    "twin_written=false · draft_written=false · live_execution_authorized=false",
    "production_router_verdict=REJECT",
  ];

  const highlight_surface = composeReadingHighlightFloatTwinFeed({
    ...input.highlight_surface,
    operator_ack: input.operator_ack,
  });
  notes.push(...highlight_surface.notes.map((n) => `[highlight_surface] ${n}`));

  const draft_pack = composeDraftBeforeMergeMoPriceCeilingRecursiveTwin({
    ...input.draft_pack,
    operator_ack: input.operator_ack,
  });
  notes.push(...draft_pack.notes.map((n) => `[draft_pack] ${n}`));

  const session_id = requireNonEmpty(
    highlight_surface.session_id,
    "session_id",
  );
  const parent_asset_id = requireNonEmpty(
    input.highlight_surface.parent_asset_id,
    "parent_asset_id",
  );
  const week_id = requireNonEmpty(draft_pack.week_id, "week_id");
  const asset_id = requireNonEmpty(draft_pack.asset_id, "asset_id");
  const title = requireNonEmpty(draft_pack.title, "title");
  const account_id = requireNonEmpty(draft_pack.account_id, "account_id");

  const session_aligned = draft_pack.session_id === session_id;
  const parent_aligned =
    draft_pack.parent_asset_id === parent_asset_id ||
    draft_pack.asset_id === parent_asset_id;
  if (!session_aligned) {
    notes.push(
      "session_id mismatch between highlight_surface and draft_pack — pack_ready blocked",
    );
  } else {
    notes.push("session_aligned=true");
  }
  if (!parent_aligned) {
    notes.push(
      "parent_asset_id mismatch between highlight_surface and draft_pack — pack_ready blocked",
    );
  } else {
    notes.push("parent_aligned=true");
  }

  let pack_ready = false;
  if (require_both) {
    pack_ready =
      session_aligned &&
      parent_aligned &&
      highlight_surface.pack_ready === true &&
      draft_pack.pack_ready === true &&
      draft_pack.production_router_verdict === "REJECT" &&
      highlight_surface.live_dispatched === false &&
      highlight_surface.merge_executed === false &&
      highlight_surface.pack_dispatched === false &&
      highlight_surface.twin_written === false &&
      draft_pack.draft_written === false &&
      draft_pack.merge_executed === false &&
      draft_pack.live_dispatched === false &&
      draft_pack.live_execution_authorized === false &&
      draft_pack.charge_executed === false &&
      draft_pack.twin_written === false &&
      draft_pack.remote_index_queried === false &&
      draft_pack.pdf_primary === false &&
      draft_pack.live_router_authorized === false &&
      draft_pack.secrets_stored === false &&
      input.operator_ack === true;
  } else {
    pack_ready =
      session_aligned &&
      parent_aligned &&
      input.operator_ack === true &&
      draft_pack.production_router_verdict === "REJECT" &&
      draft_pack.pdf_primary === false &&
      highlight_surface.live_dispatched === false &&
      (highlight_surface.pack_ready === true || draft_pack.pack_ready === true);
  }

  if (pack_ready) {
    notes.push(
      "pack_ready=true — floating DR highlight + draft-before-merge MO price-ceiling ready; still pure",
    );
  } else {
    notes.push(
      "pack_ready=false — highlight_surface, draft_pack, alignment, or operator_ack gate open",
    );
  }

  if (
    highlight_surface.live_dispatched !== false ||
    highlight_surface.merge_executed !== false ||
    highlight_surface.pack_dispatched !== false ||
    highlight_surface.twin_written !== false ||
    draft_pack.draft_written !== false ||
    draft_pack.merge_executed !== false ||
    draft_pack.live_dispatched !== false ||
    draft_pack.live_execution_authorized !== false ||
    draft_pack.charge_executed !== false ||
    draft_pack.twin_written !== false ||
    draft_pack.remote_index_queried !== false ||
    draft_pack.pdf_primary !== false ||
    draft_pack.live_router_authorized !== false ||
    draft_pack.secrets_stored !== false ||
    draft_pack.production_router_verdict !== "REJECT"
  ) {
    throw new Error("invariant: honesty flags must remain false / REJECT");
  }

  notes.push("live_dispatched=false");
  notes.push("merge_executed=false");
  notes.push("pack_dispatched=false");
  notes.push("twin_written=false");
  notes.push("record_persisted=false");
  notes.push("draft_written=false");
  notes.push("live_execution_authorized=false");
  notes.push("charge_executed=false");
  notes.push("prompts_injected=false");
  notes.push("live_router_authorized=false");
  notes.push("secrets_stored=false");
  notes.push("live_meter_read=false");
  notes.push("remote_index_queried=false");
  notes.push("backlog_mutated=false");
  notes.push("store_mutated=false");
  notes.push("suite_rewritten=false");
  notes.push("pdf_view_authorized=false");
  notes.push("pdf_primary=false");
  notes.push("live_dispatch_authorized=false");
  notes.push("inventory_mutated=false");
  notes.push("analysis_written=false");
  notes.push("purchase_executed=false");
  notes.push("hosted=false");
  notes.push("remote_fetched=false");
  notes.push("production_router_verdict=REJECT");

  return {
    session_id,
    parent_asset_id,
    week_id,
    asset_id,
    title,
    account_id,
    highlight_surface,
    draft_pack,
    session_aligned,
    parent_aligned,
    pack_ready,
    live_dispatched: false,
    merge_executed: false,
    pack_dispatched: false,
    twin_written: false,
    record_persisted: false,
    draft_written: false,
    live_execution_authorized: false,
    charge_executed: false,
    prompts_injected: false,
    live_router_authorized: false,
    secrets_stored: false,
    live_meter_read: false,
    remote_index_queried: false,
    backlog_mutated: false,
    store_mutated: false,
    suite_rewritten: false,
    pdf_view_authorized: false,
    pdf_primary: false,
    live_dispatch_authorized: false,
    inventory_mutated: false,
    analysis_written: false,
    purchase_executed: false,
    hosted: false,
    remote_fetched: false,
    production_router_verdict: "REJECT",
    notes,
    authority:
      "floating_dr_draft_before_merge_mo_price_ceiling_compose_advisory",
  };
}

export function formatFloatingDrDraftBeforeMergeMoPriceCeilingSummary(
  c: FloatingDrDraftBeforeMergeMoPriceCeilingCompose,
): string {
  return (
    `pack_ready=${c.pack_ready} · ` +
    `highlight_ready=${c.highlight_surface.pack_ready} · ` +
    `draft_ready=${c.draft_pack.pack_ready} · ` +
    `session_aligned=${c.session_aligned} · ` +
    `parent_aligned=${c.parent_aligned} · ` +
    `verdict=${c.production_router_verdict} · ` +
    `live_dispatched=false · twin_written=false · draft_written=false · ` +
    `live_execution_authorized=false`
  );
}
